src/libmongoc/src/mongoc/mongoc-cluster.c

```diff
@@ -641,6 +641,53 @@ mongoc_cluster_run_command_monitored (mongoc_cluster_t *cluster,
    return retval;
 }
 
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_should_use_op_msg --
+ *
+ * Returns:
+ *       true if op_msg should be used to transport the given command
+ *
+ * Side effects:
+ *      none
+ *
+ *--------------------------------------------------------------------------
+ */
+static bool _mongoc_should_use_op_msg(const mongoc_cluster_t *cluster, const mongoc_cmd_t *cmd)
+{
+ /* If we're running a newer wire protocol OR have manually selected a server protocol,
+ always select OP_MSG: */
+
+ return (WIRE_VERSION_OP_MSG <= cmd->server_stream->sd->max_wire_version) || 
+        mongoc_client_uses_server_api(cluster->client);
```

Comment:
Is the intended change is to use OP_MSG with `hello` on the first command if a server API version is set as described in CDRIVER-4193?

If so, [mongoc-async-cmd.c](https://github.com/mongodb/mongo-c-driver/commit/aa76669d60e3bdc9670eedcf64bc12472bd03ec8#diff-d35f0d63a4ad8bfb630b204735bbc00dab445deed4a03e0e23793671c5fb4152R26) will need to be updated. That is used by a single threaded `mongoc_client_t` for scanning servers. This can be tested with a mock server test (noted below).

Reply:
Yes, that's right. Basically, the only time we should now speak OP_QUERY is with an old wire version on the server and no specified API (which would imply the higher version anyway). I see _mongoc_async_cmd_init_send() in mongoc-async-cmd.c, I'll see what I can do.

Reply:
As discussed offline, this PR reverts async work.

Reply:
Ah, my misunderstanding! I'll revisit changes to mongo-async-cmd.c; getting the information about the requested API into that layer may require changing some of the signatures, I will look again but last time I didn't see the client settings exposed to them.